src/main/java/org/elasticsearch/cluster/ClusterState.java

```diff
@@ -642,5 +688,100 @@ public static ClusterState readFrom(StreamInput in, @Nullable DiscoveryNode loca
             }
             return builder.build();
         }
+
+    }
+
+    @Override
+    public Diff diff(ClusterState previousState) {
+        return new ClusterStateDiff(previousState, this);
+    }
+
+    @Override
+    public ClusterState readFrom(StreamInput in) throws IOException {
+        return Builder.readFrom(in, null);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        Builder.writeTo(this, out);
+    }
+
+    private static class ClusterStateDiff implements Diff {
+
+        private final long version;
+
+        private final String previousUuid;
+
+        private final String uuid;
+
+        private final ClusterName clusterName;
+
+        private final Diff routingTable;
+
+        private final Diff nodes;
+
+        private final Diff metaData;
+
+        private final Diff blocks;
+
+        private final Diff customs;
+
+        public ClusterStateDiff(ClusterState before, ClusterState after) {
+            previousUuid = before.uuid;
+            uuid = after.uuid;
+            version = after.version;
+            clusterName = after.clusterName;
+            routingTable = after.routingTable.diff(before.routingTable);
+            nodes = after.nodes.diff(before.nodes);
+            metaData = after.metaData.diff(before.metaData);
+            blocks = after.blocks.diff(before.blocks);
+            customs = DiffableUtils.diff(before.customs, after.customs);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            clusterName.writeTo(out);
+            out.writeString(previousUuid);
+            out.writeString(uuid);
+            out.writeLong(version);
+            routingTable.writeTo(out);
+            nodes.writeTo(out);
+            metaData.writeTo(out);
+            blocks.writeTo(out);
+            customs.writeTo(out);
+        }
+
+        public static ClusterState readDiffAndApply(ClusterState state, StreamInput in)  throws IOException {
+            Builder builder = new Builder(ClusterName.readClusterName(in));
+            String previousUuid = in.readString();
+            String uuid = in.readString();
+            long version = in.readLong();
+            if (uuid.equals(state.uuid)) {
```

Comment:
This can occur if a non-master node requests to send complete cluster state while master publishes a new cluster state and if a complete cluster state arrives just before the diff for it.